xds/src/main/java/io/grpc/xds/FilterChainMatchingHandler.java

```diff
@@ -0,0 +1,362 @@
+          InetAddress cidrAddr = cidrRange.getAddressPrefix();
+          boolean cidrIsIpv6 = cidrAddr instanceof Inet6Address;
+          if (isIPv6 == cidrIsIpv6) {
+            int prefixLen = cidrRange.getPrefixLen();
+            CidrMatcher matcher = CidrMatcher.create(cidrAddr, prefixLen);
+            if (matcher.matches(address) && prefixLen > matchingPrefixLength) {
+              matchingPrefixLength = prefixLen;
+            }
+          }
+        }
+      }
+      return matchingPrefixLength;
+    }
+
+    // use prefix_ranges (CIDR) and get the most specific matches
+    private static Collection<FilterChain> filterOnIpAddress(
+            Collection<FilterChain> filterChains, InetAddress address, boolean forDestination) {
+      // curent list of top ones
+      ArrayList<FilterChain> topOnes = new ArrayList<>(filterChains.size());
+      int topMatchingPrefixLen = -1;
+      for (FilterChain filterChain : filterChains) {
+        int currentMatchingPrefixLen =
+            getMatchingPrefixLength(filterChain.getFilterChainMatch(), address, forDestination);
+
+        if (currentMatchingPrefixLen >= 0) {
+          if (currentMatchingPrefixLen < topMatchingPrefixLen) {
+            continue;
+          }
+          if (currentMatchingPrefixLen > topMatchingPrefixLen) {
+            topMatchingPrefixLen = currentMatchingPrefixLen;
+            topOnes.clear();
+          }
+          topOnes.add(filterChain);
+        }
+      }
+      return topOnes;
+    }
+  }
+
+  static class FilterChainMatchingNegotiatorServerFactory
+          implements InternalProtocolNegotiator.ServerFactory {
+    private final InternalProtocolNegotiator.ServerFactory delegate;
+
+    public FilterChainMatchingNegotiatorServerFactory(
+            InternalProtocolNegotiator.ServerFactory delegate) {
+      this.delegate = checkNotNull(delegate, "delegate");
+    }
+
+    @Override
+    public ProtocolNegotiator newNegotiator(
+            final ObjectPool<? extends Executor> offloadExecutorPool) {
+
+      class FilterChainMatchingNegotiator implements ProtocolNegotiator {
+
+        @Override
+        public AsciiString scheme() {
+          return SCHEME;
+        }
+
+        @Override
+        public ChannelHandler newHandler(GrpcHttp2ConnectionHandler grpcHandler) {
+          AtomicReference<FilterChainSelector> filterChainSelectorRef =
+                  grpcHandler.getEagAttributes().get(ATTR_FILTER_CHAIN_SELECTOR_REF);
+          checkNotNull(filterChainSelectorRef, "filterChainSelectorRef");
+          return new FilterChainMatchingHandler(grpcHandler, filterChainSelectorRef.get(),
+                  delegate.newNegotiator(offloadExecutorPool));
+        }
+
+        @Override
+        public void close() {
+        }
+      }
+
+      return new FilterChainMatchingNegotiator();
+    }
+  }
+
+  /**
+   * The FilterChain level configuration.
+   */
+  public static final class SelectedConfig {
```

Comment:
Doesn't need to be public

Reply:
If you do end up making `SelectedConfig` as the interface between `FilterChainMatchingHandler` and the delegate (or downstream) handler then this would indeed need to be public.

Reply:
This is what I am planning to extend to. For now it can be private.